pkg/peaks/peaks.go

```diff
@@ -0,0 +1,220 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peaks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+
+	"github.com/paypal/load-watcher/pkg/watcher"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"sigs.k8s.io/scheduler-plugins/apis/config"
+	cfgv1 "sigs.k8s.io/scheduler-plugins/apis/config/v1"
+	"sigs.k8s.io/scheduler-plugins/pkg/trimaran"
+)
+
+const (
+	Name = "Peaks"
+)
+
+type Peaks struct {
+	handle    framework.Handle
+	collector *trimaran.Collector
+	args      *config.PeaksArgs
+}
+
+type PowerModel struct {
+	K0 float64 `json:"k0"`
+	K1 float64 `json:"k1"`
+	K2 float64 `json:"k2"`
+	// Power = K0 + K1 * e ^(K2 * x) : where x is utilisation
+}
+
+var _ framework.ScorePlugin = &Peaks{}
+var maxPower = 0.0
+var clusterPowerModel map[string]PowerModel
+
+func (pl *Peaks) Name() string {
+	return Name
+}
+
+func initNodePowerModels() error {
+	data, err := os.ReadFile("/power_model/node_power_model")
+	if err != nil {
+		klog.ErrorS(err, "Unable to read power model from configMap")
+		return err
+	}
+	if err = json.Unmarshal(data, &clusterPowerModel); err != nil {
+		klog.ErrorS(err, "Unable to unmarshal power model from configMap value")
+		return err
+	}
+	return nil
+}
+
+func New(_ context.Context, obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	klog.V(4).InfoS("Peaks plugin Input config %+v\n", obj)
+
+	args, ok := obj.(*config.PeaksArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type PeaksArgs, got %T", obj)
+	}
+	collector, err := trimaran.NewCollector(&config.TrimaranSpec{WatcherAddress: args.WatcherAddress})
+	if err != nil {
+		return nil, err
+	}
+	err = initNodePowerModels()
+	if err != nil {
+		return nil, err
+	}
+	pl := &Peaks{
+		handle:    handle,
+		collector: collector,
+		args:      args,
+	}
+	return pl, nil
+}
+
+func (pl *Peaks) Score(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	score := framework.MinNodeScore
+
+	nodeInfo, err := pl.handle.SnapshotSharedLister().NodeInfos().Get(nodeName)
+	if err != nil {
+		return score, framework.NewStatus(framework.Error, fmt.Sprintf("getting node %q from Snapshot: %v", nodeName, err))
+	}
+
+	metrics, _ := pl.collector.GetNodeMetrics(nodeName)
+	if metrics == nil {
+		klog.ErrorS(nil, "Failed to get metrics for node; using minimum score", "nodeName", nodeName)
+		return score, nil
+	}
+
+	var curPodCPUUsage int64
+	for _, container := range pod.Spec.Containers {
+		curPodCPUUsage += PredictUtilisation(&container)
+	}
+	if pod.Spec.Overhead != nil {
+		curPodCPUUsage += pod.Spec.Overhead.Cpu().MilliValue()
+	}
```

Comment:
There is some corner case not being considered here: like initContainers and potential k8s alpha feature like `InPlacePodVerticalScalingEnabled`. I'd recommend to re-use existing k8s utility to calculate a pod's CPU request so it's adaptive to future k8s features:

```go
import (
    "k8s.io/kubernetes/pkg/api/v1/resource"
)

reqs :=  resource.PodRequests(pod, resource.PodResourcesOptions{})
// then fetch the CPU part
```

Reply:
Can we fix this through a separate issue after PR is merged?